datashape/discovery.py

```diff
@@ -396,3 +396,29 @@ def descendents(d, x):
         children -= desc
         desc.update(children)
     return desc
+
+
+def _tp_eq(a, b):
+    return a == b or b is null
```

Comment:
This is probably a bad name. The idea is that `null` represents an an infinite bound, or default step, so this was used to check that all the types for start, stop, and step are the same or null.